### PR TITLE
separate iavl pruning from asyncCommitRoutine

### PR DESCRIPTION
**Pull request by cwbhhjl**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: libs/iavl/mutable_tree.go [hunk 1]
```diff
@@ -900,6 +902,16 @@ func (tree *MutableTree) SaveVersionSync(version int64, useDeltas bool) ([]byte,
 }
 
 func (tree *MutableTree) deleteVersion(batch dbm.Batch, version int64, versions map[int64]bool) error {
+	if err := tree.deleteVersionPrecheck(version, versions); err != nil {
+		return err
+	}
+
+	tree.ndb.deleteVersion(batch, version, true)
```
Comment (ilovers) [inline libs/iavl/mutable_tree.go:909]: why is error checking ignored in line?

**Review by ilovers — DISMISSED** [review]

File: libs/iavl/mutable_tree_oec.go [hunk 2]
```diff
@@ -259,6 +269,35 @@ func (tree *MutableTree) commitSchedule() {
 		}
 	}
 }
+
+func (tree *MutableTree) pruningSchedule() {
+	for event := range tree.pruneCh {
+		if event.version >= 0 {
+			trc := trace.NewTracer("pruningSchedule")
+			trc.Pin("Pruning")
```
Comment (giskook) [inline libs/iavl/mutable_tree_oec.go]: Suggest trace both deleteVersion and commit.

File: libs/iavl/mutable_tree.go [hunk 3]
```diff
@@ -900,6 +902,16 @@ func (tree *MutableTree) SaveVersionSync(version int64, useDeltas bool) ([]byte,
...
+
+	return nil
+}
+
+func (tree *MutableTree) deleteVersionPrecheck(version int64, versions map[int64]bool) error {
```
Comment (giskook) [inline libs/iavl/mutable_tree.go]:
```suggestion
func (tree *MutableTree) deleteVersionPreCheck(version int64, versions map[int64]bool) error {
```

File: libs/iavl/mutable_tree.go [hunk 4]
```diff
@@ -114,6 +115,7 @@ func NewMutableTreeWithOpts(db dbm.DB, cacheSize int, opts *Options) (*MutableTr
 			historyStateNum:      MaxCommittedHeightNum,
 
 			commitCh:          make(chan commitEvent),
+			pruneCh:           make(chan pruneEvent, 10),
```
Comment (giskook) [inline libs/iavl/mutable_tree.go]: use constant instead of 10

File: libs/iavl/mutable_tree_oec.go [hunk 5]
```diff
@@ -337,20 +379,13 @@ func (tree *MutableTree) updateCommittedStateHeightPool(batch dbm.Batch, version
 		delete(tree.committedHeightMap, oldVersion)
 
 		if EnablePruningHistoryState {
-			if writeToDB {
-				batch = tree.ndb.db.NewBatch()
-			}
-			if err := tree.deleteVersion(batch, oldVersion, versions); err != nil {
+
```
Comment (giskook) [inline libs/iavl/mutable_tree_oec.go]: remove the blank line

File: libs/iavl/mutable_tree_oec.go [hunk 6]
```diff
@@ -164,15 +169,20 @@ func (tree *MutableTree) setNewWorkingTree(version int64, persisted bool) ([]byt
 	}
 	treeMap.updatePpnc(version)
 
+	tree.updatePruningVersion()
```
Comment (giskook) [inline libs/iavl/mutable_tree_oec.go:173]: It's a little different with previous logical. now every height will prune?  If so the `historyStateNum` will unuseless? I think we should be careful here.

Comment (cwbhhjl) [reply libs/iavl/mutable_tree_oec.go:173]: The logic of prune is still controlled by ac and historyStateNum

**Review by ilovers — APPROVED** [review]

File: libs/iavl/mutable_tree.go [hunk 7]
```diff
@@ -114,6 +115,7 @@ func NewMutableTreeWithOpts(db dbm.DB, cacheSize int, opts *Options) (*MutableTr
 			historyStateNum:      MaxCommittedHeightNum,
 
 			commitCh:          make(chan commitEvent),
+			pruneCh:           make(chan pruneEvent, PruningChannelSize),
```
Comment (giskook) [inline libs/iavl/mutable_tree.go:118]: Can the pruneCh full? if the channel is full what will happen?

Comment (cwbhhjl) [reply libs/iavl/mutable_tree.go:118]: If each prune takes a lot of time, then the pruneCh will fill up. If the pruneCh is full, the program will hang until a prune is completed by the pruneRoutine

**Review by giskook — APPROVED** [review]

**Pull request merged**